File: app/src/main/java/nich/work/aequorea/common/di/ViewModelProviderFactory.java
package nich.work.aequorea.common.di;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

/**
 * A provider factory that persists ViewModels{@link ViewModel}.
 * Used if the viewmodel has a parameterized constructor.
 */
public class ViewModelProviderFactory<V> implements ViewModelProvider.Factory {
    
    private V viewModel;
    
    public ViewModelProviderFactory(V viewModel) {
        this.viewModel = viewModel;
    }
    
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(viewModel.getClass())) {
            return (T) viewModel;
        }
        throw new IllegalArgumentException("Unknown class name");
    }
}

File: app/src/main/java/nich/work/aequorea/common/di/ViewModelProviderFactory.kt
package nich.work.aequorea.common.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

/**
 * A provider factory that persists ViewModels[ViewModel].
 * Used if the viewModel has a parameterized constructor.
 */
class ViewModelProviderFactory(private val viewModel: Any) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(viewModel::class.java)) {
            return viewModel as T
        }
        throw IllegalArgumentException("Unknown class name")
    }

}
